ProjFS.Mac/PrjFSKextTests/VnodeCacheTests.mm

```diff
@@ -0,0 +1,700 @@
+    shared_ptr<vnode> testVnodeFile2;
+    shared_ptr<vnode> testVnodeFile3;
+    PerfTracer dummyPerfTracer;
+    vfs_context_t dummyVFSContext;
+}
+
+static const VirtualizationRootHandle TestRootHandle = 51;
+static const VirtualizationRootHandle TestSecondRootHandle = 52;
+
+// Helper class for interacting with s_entries and s_entriesCapacity
+class VnodeCacheEntriesWrapper
+{
+public:
+    VnodeCacheEntriesWrapper(const uint32_t capacity, const bool fillCache)
+    {
+        s_entriesCapacity = capacity;
+        s_entries = new VnodeCacheEntry[s_entriesCapacity];
+        
+        static shared_ptr<mount> dummyMount = mount::Create();
+        static shared_ptr<vnode> dummyNode = dummyMount->CreateVnodeTree("/DUMMY");
+        for (uint32_t i = 0; i < s_entriesCapacity; ++i)
+        {
+            if (fillCache)
+            {
+                s_entries[i].vnode = dummyNode.get();
+            }
+            else
+            {
+                memset(&(s_entries[i]), 0, sizeof(VnodeCacheEntry));
+            }
+        }
+    }
+    
+    ~VnodeCacheEntriesWrapper()
+    {
+        s_entriesCapacity = 0;
+        delete[] s_entries;
+    }
+    
+    void MarkEntryAsFree(const uintptr_t entryIndex)
+    {
+        s_entries[entryIndex].vnode = nullptr;
+    }
+    
+    VnodeCacheEntry& operator[] (const uintptr_t entryIndex)
+    {
+        return s_entries[entryIndex];
+    }
+    
+    uint32_t GetCapacity() const
+    {
+        return s_entriesCapacity;
+    }
+};
+
+- (void) setUp
+{
+    kern_return_t initResult = VirtualizationRoots_Init();
+    XCTAssertEqual(initResult, KERN_SUCCESS);
+    
+    self->testMount = mount::Create();
+    repoPath = "/Users/test/code/Repo";
+    self->repoRootVnode = self->testMount->CreateVnodeTree(repoPath, VDIR);
+    self->testVnodeFile1 = testMount->CreateVnodeTree(repoPath + "/file1");
+    self->testVnodeFile2 = testMount->CreateVnodeTree(repoPath + "/file2");
+    self->testVnodeFile3 = testMount->CreateVnodeTree(repoPath + "/file3");
+    self->dummyVFSContext = vfs_context_create(nullptr);
+}
+
+- (void) tearDown
+{
+    vfs_context_rele(self->dummyVFSContext);
+    self->testVnodeFile3.reset();
+    self->testVnodeFile2.reset();
+    self->testVnodeFile1.reset();
+    self->repoRootVnode.reset();
+    self->testMount.reset();
+    MockCalls::Clear();
+    VirtualizationRoots_Cleanup();
+}
+
+- (void)testVnodeCache_FindRootForVnode_EmptyCache {
+    VnodeCacheEntriesWrapper cacheWrapper(/* capacity*/ 100, /* fillCache*/ false);
+    
+    VirtualizationRootHandle repoRootHandle = InsertVirtualizationRoot_Locked(
+        nullptr /* no client */,
+        0,
+        self->repoRootVnode.get(),
+        self->repoRootVnode->GetVid(),
+        FsidInode{ self->repoRootVnode->GetMountPoint()->GetFsid(), self->repoRootVnode->GetInode() },
+        self->repoPath.c_str());
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(repoRootHandle));
+    
+    MockCalls::Clear();
```

Comment:
I will add some comments in the code.

> Why do we do the clear after the first functions?

I'm clearing `MockCalls` here because we're interested in what function calls were made by the vnode cache code, and the function calls made by the code for setting up the test.

> Which funcitons do we want to ensure aren't called at the end?

This is primarily a sanity check.  We don't expect the vnode cache to be hitting _any_ of our mocked out functions.